### DISCUSSION TITLE
Voronoï diagram corresponding to constrained Delaunay triangulation

### DISCUSSION THREAD

**Discussion by stla** [CGAL/cgal#7303] [Q&A] [answered]
Hello,

A couple of years ago, I did a [R package](https://stla.github.io/tessellation/articles/the-tessellation-package.html) for Delaunay and Voronoï tessellations. Now I'd like to implement the Voronoï tessellation dual to a 2D constrained Delaunay triangulation.

- Am I right to believe that CGAL is only able to compute the Voronoï diagram for an unconstrained Delaunay triangulation? I took a look at the doc, and I understood that it is computed from a given set of sites, so my guess is that the unconstrained Delaunay triangulation of these sites is constructed.

- If yes, I'll have to implement the construction from the constrained Delaunay triangulation. Would you know if the principle of the construction is the same as the unconstrained case? Or is there any difference?

**Comment by MaelRL** [comment, marked as answer]
The class [`Voronoi_diagram_2`](https://doc.cgal.org/latest/Voronoi_diagram_2/classCGAL_1_1Voronoi__diagram__2.html) is a simple wrapper around a Delaunay triangulation.

It does not compute anything, it only access combinatorial and geometric information of the primal graph and converts it on the fly to a Voronoi representation: for example, the Voronoi vertex `v1` adjacent to a Voronoi vertex `v0` is the dual of the adjacent face `f1` adjacent to `f0`, the face dual of `v0`.

As such, there shouldn't be anything that prevents using it with a CDT, but you probably need to define the adaptation helper classes (see concepts `AdaptationTraits_2`, `AdaptationPolicy_2`).

You can find some examples of implementations for different Delaunay graphs in the `include` directory of the Voronoi package (https://github.com/CGAL/cgal/tree/master/Voronoi_diagram_2/include/CGAL). [Here](https://github.com/CGAL/cgal/blob/master/Voronoi_diagram_2/include/CGAL/Delaunay_triangulation_on_sphere_adaptation_traits_2.h) is for example the adaptation traits to construct Voronoi diagrams of the [Triangulations on sphere package](https://doc.cgal.org/latest/Triangulation_on_sphere_2/index.html#Chapter_2D_Triangulations_on_sphere).

**Reply by stla** [reply]
Thanks. That sounds complicated to me. I will "manually" construct the Voronoï diagram from the constrained Delaunay triangulation (Joseph O'Rourke confirmed it's the same as the unconstrained case).

**Comment by stla** [comment]
I get it. But it remains to truncate the Voronoi edges that go outside the Delaunay triangulation...

![Rplot](https://user-images.githubusercontent.com/4466543/222910708-516d989f-58cb-4982-8dfc-6ffa3987af41.png)

**Reply by MaelRL** [reply]
> I get it. But it remains to truncate the Voronoi edges that go outside the Delaunay triangulation...
> 
> ![Rplot](https://user-images.githubusercontent.com/4466543/222910708-516d989f-58cb-4982-8dfc-6ffa3987af41.png)

You can use [`mark_domain_in_triangulation()`](https://cgal.geometryfactory.com/CGAL/doc/master/Triangulation_2/group__PkgTriangulation2Miscellaneous.html#ga0409755d0eb89100810230443a85e7eb), and not draw edges for which at least one incident face of the Delaunay graph is not in the domain.

**Reply by MaelRL** [reply]
Ah, no, you're already doing that and you want to truncate Voronoi edges. Then you can use `CDT::locate()` with the Voronoi vertices in the triangulation and if it finds a face that is not in the domain, you have to clip (or even fully remove) the edge.

**Reply by stla** [reply]
Thanks. Fine for locating. But I don't see yet how to clip. Is there some tool in CGAL that could help? The mesh clipping would not help, because it is restricted to triangle meshes.

**Reply by stla** [reply]
Maybe with the polygons intersection...

**Reply by stla** [reply]
I see, I can paint the inner hexagon in white (easy) and the complementary of the outer hexagon in white (using the CGAL features for polygons).

**Comment by stla** [comment]
I'm not sure this is a constrained Voronoï diagram... because the constraining edges are edges of the Delaunay triangulation here. Shouldn't they be edges of some Voronoï cells rather? This is my impression when I look at this picture:

![](https://i.stack.imgur.com/kRu3m.png)

So I mean a constrained Voronoï diagram is *not* a Voronoï diagram of a constrained Delaunay triangulation... What do you think? Will see if I find a definition somewhere.

**Reply by MaelRL** [reply]
https://hal.inria.fr/inria-00523812/file/centroidal-vd.pdf is probably of interest. @janetournois will know if/where there is some code (in CGAL or somewhere else).

**Reply by stla** [reply]
I'm right: [this paper](https://link.springer.com/article/10.1007/BF01188709) claims that the constrained Voronoi diagram is *not* the dual of the constrained Delaunay triangulation.

So I think the constrained Voronoï diagram is some job for CGAL ;-)

Thanks for the link, will take a look.

**Comment by stla** [comment]
![Rplot2](https://user-images.githubusercontent.com/4466543/223277363-892f6eb9-5808-4f0c-bc90-cac102540cd2.png)

**Comment by janetournois** [comment]
Hello,
indeed the "Bounded Voronoi diagram" described by Seidel is implemented in CGAL as an implementation detail of [lloyd_optimize_mesh_2()](https://doc.cgal.org/latest/Mesh_2/group__PkgMesh2Functions.html#ga5903d196ab84ecf1c1596c8892f7a6d4).

The full construction of the BVD is described in [this paper](http://www.global-sci.org/v1/nmtma/volumes/v3n2/pdf/32-212.pdf?code=29Crx%2BQf1001lsLVT2Hegg%3D%3D) and the main header is [CGAL/Constrained_voronoi_diagram_2.h](https://github.com/CGAL/cgal/blob/master/Mesh_2/include/CGAL/Constrained_voronoi_diagram_2.h)

**Reply by stla** [reply]
Thanks. Unfortunately understanding this code is beyond my skills. 

I'm happy with my implementation, except I don't know how to clip the Voronoï cells that go outside the constraining region.

![voronoiDiagram2](https://user-images.githubusercontent.com/4466543/223700926-237a74ce-1dda-4639-8142-effbcc5c154d.png)
